utils/ssh_client.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import logging
+import os
+import time
+from collections import namedtuple
+
+import paramiko
+
+logger = logging.getLogger(__name__)
+CMDResult = namedtuple('CMDResult', ['stdout', 'stderr', 'r_code', 'exec_time'])
+
+
+class SSHClient(object):
+
+    def __init__(self,
+                 host=None,
+                 username=None,
+                 password=None,
+                 key=None,
+                 passphrase=None,
+                 allow_unknown_hosts=False,
+                 environment=None,
+                 use_credentials=False):
+
+        self.host = host
+        self.username = username
+        self.password = password
+        self.key = key
+        self.passphrase = passphrase
+        self.allow_unknown_hosts = allow_unknown_hosts
+        self.environment = environment
+        self.client = paramiko.SSHClient()
+        self.use_credentials = use_credentials
+
+    def __del__(self):
+        self.client.close()
+        logger.info('SSH client was closed')
+
+    @property
+    def connected(self):
+        return False if self.client.get_transport() is None else True
+
+    def load_system_known_hosts(self, path=None):
+        if path is None:
+            self.client.load_host_keys(os.path.expanduser('~/.ssh/known_hosts'))
+        else:
+            self.client.load_host_keys(path)
+
+    def clear_host_keys(self):
+        self.client.get_host_keys().clear()
+
+    def _connect_with_password(self, **kwargs):
+        if self.password is None:
+            message = 'Password is not specified for {}@{}'.format(self.username, self.host)
+            logger.error(message)
+            raise SSHConnectionFailed(message)
+        self.client.connect(self.host, username=self.username, password=self.password, **kwargs)
+
+    def _connect_with_key(self, **kwargs):
+        if self.key is None:
+            message = 'Private key is not specified for {}@{}. Attempt to connect using credentials'.format(
+                self.username, self.host)
+            logger.warning(message)
+            raise SSHKeyMissing(message)
+        if self.passphrase is None:
+            message = 'Passphrase is not specified for {}'.format(self.key)
+            logger.error(message)
+            raise SSHConnectionFailed(message)
+        self.client.connect(self.host, username=self.username, key_filename=self.key, passphrase=self.passphrase,
+                            **kwargs)
+
+    def connect(self, use_credentials=False, **kwargs):
```

Comment:
Why this method doesn't use default value for ```use_credentials``` which was set on class initialization?

Reply:
Done